src/libclipper/src/task_executor.cpp

```diff
@@ -59,21 +62,68 @@ void PredictionCache::put(const VersionedModelId &model,
                           const Output &output) {
   std::unique_lock<std::mutex> l(m_);
   auto key = hash(model, input->hash());
-  auto search = cache_.find(key);
-  if (search != cache_.end()) {
-    if (!search->second.completed_) {
+  auto search = entries_.find(key);
+  if (search != entries_.end()) {
+    CacheEntry &entry = search->second;
+    if (!entry.completed_) {
       // Complete the outstanding promises
-      for (auto &p : search->second.value_promises_) {
+      for (auto &p : entry.value_promises_) {
         p.set_value(output);
       }
-      search->second.completed_ = true;
-      search->second.value_ = output;
+      entry.completed_ = true;
+      entry.value_ = output;
+      size_bytes_ += output.y_hat_.size();
+      evict_entries(size_bytes_ - max_size_bytes_);
     }
   } else {
     CacheEntry new_entry;
     new_entry.value_ = output;
     new_entry.completed_ = true;
-    cache_.insert(std::make_pair(key, std::move(new_entry)));
+    insert_entry(key, new_entry);
+  }
+}
+
+void PredictionCache::insert_entry(const long key, CacheEntry &value) {
+  size_t entry_size_bytes = value.completed_ ? value.value_.y_hat_.size() : 0;
+  if (entry_size_bytes <= max_size_bytes_) {
+    evict_entries(size_bytes_ + entry_size_bytes - max_size_bytes_);
+    page_buffer_.insert(page_buffer_.begin() + page_buffer_index_, key);
+    page_buffer_index_ = (page_buffer_index_ + 1) % page_buffer_.size();
+    size_bytes_ += entry_size_bytes;
+    entries_.insert(std::make_pair(key, std::move(value)));
+  } else {
+    // This entry is too large to cache
+    log_error_formatted(LOGGING_TAG_TASK_EXECUTOR,
+                        "Received an output of size: {} bytes that exceeds "
+                        "cache size of: {} bytes",
+                        entry_size_bytes, max_size_bytes_);
+  }
+}
+
+void PredictionCache::evict_entries(long space_needed_bytes) {
+  if (space_needed_bytes <= 0) {
+    return;
+  }
+  while (space_needed_bytes > 0) {
+    long page_key = page_buffer_[page_buffer_index_];
```

Comment:
See the above comment. We use `page_buffer_` and `page_buffer_index_` to evict as many pages as necessary to obtain the required free space.